lib/wasi/src/fs/inode_guard.rs

```diff
@@ -237,209 +198,229 @@ impl<'a> Future for InodeValFilePollGuardJoin<'a> {
+                                    },
                                 },
                             },
+                            Eventtype::Clock => EventUnion { clock: 0 },
                         },
-                        Eventtype::Clock => EventUnion { clock: 0 },
-                    },
-                });
-            }
+                    })
+                    .ok();
+                }
+                Poll::Pending => {}
+            };
         }
         if has_write {
-            let mut poll_result = match &mut self.mode {
+            let poll_result = match &mut self.mode {
                 InodeValFilePollGuardMode::File(file) => {
                     let mut guard = file.write().unwrap();
                     let file = Pin::new(guard.as_mut());
                     file.poll_write_ready(cx)
                 }
-                InodeValFilePollGuardMode::EventNotifications {
-                    waker,
-                    counter,
-                    immediate,
-                    ..
-                } => {
-                    if *immediate {
-                        let cnt = counter.load(Ordering::Acquire);
-                        Poll::Ready(Ok(cnt as usize))
+                InodeValFilePollGuardMode::EventNotifications(inner) => {
+                    {
+                        let mut guard = inner.wakers.lock().unwrap();
+                        if guard.iter().any(|a| a.will_wake(waker)) == false {
+                            guard.push_front(waker.clone());
+                        }
+                    }
+                    let val = inner.counter.load(Ordering::Acquire);
+                    if inner.last_poll.swap(val, Ordering::AcqRel) != val {
+                        Poll::Ready(Ok(val as usize))
                     } else {
-                        let counter = counter.clone();
-                        let mut waker = waker.get_mut().unwrap();
-                        let mut notifications = Pin::new(waker.deref_mut());
-                        notifications.poll_recv(cx).map(|_| {
-                            let cnt = counter.load(Ordering::Acquire);
-                            Ok(cnt as usize)
-                        })
+                        Poll::Pending
                     }
                 }
                 InodeValFilePollGuardMode::Socket { ref inner } => {
-                    let mut guard = inner.write().unwrap();
-                    guard.poll_write_ready(cx).map_err(net_error_into_io_err)
+                    let mut guard = inner.protected.write().unwrap();
+                    let res = guard.poll_write_ready(cx).map_err(net_error_into_io_err);
+                    match res {
+                        Poll::Ready(Err(err)) if is_err_closed(&err) => {
+                            tracing::trace!("socket write ready error (fd={}) - {}", fd, err);
+                            if std::mem::replace(&mut guard.notifications.closed, true) == false {
```

Comment:
Using `replace` for a boolean is pretty odd, you can just dereference and set the value,  would be more readable.

Reply:
Also, shouldn't all this code be on the guard poll_write_ready directly?

Reply:
that creates code duplication as it needs to be implemented for all socket implementations so it needs to stay here.
I prefer the std::mem::replace to be honest as its less lines of code

I made it slightly easier to read with this https://github.com/wasmerio/wasmer/commit/e82fc87a1d53f68b9a651b6afe96c3360fd30d1e

Reply:
Not a big deal either way, that's a preference thing.